llama.cpp

```diff
@@ -7328,11 +7328,115 @@ void llama_sample_top_k(struct llama_context * ctx, llama_token_data_array * can
     }
 }
 
+void read_or_write_base_min_p(float* base_min_p) {
+    // Define the filename
+    const std::string filename = "SamplerBaseMinP.txt";
+
+    // Check if the file exists
+    std::ifstream infile(filename);
+    if (!infile.good()) {
+        // File doesn't exist, create it with default values
+        std::ofstream outfile(filename);
+        if (outfile.is_open()) {
+            // Set the default value for base_min_p
+            *base_min_p = 0.05f;
+            outfile << "base_min_p = " << *base_min_p << "\n";
+            outfile.close();
+        } else {
+            // Handle the error during file opening or writing here (optional)
+            // For example, you might want to set a safe default value or notify the user
+        }
+    } else {
+        // File exists, read the values from it
+        std::string line;
+        while (getline(infile, line)) {
+            std::istringstream iss(line);
+            std::string key;
+            float value;
+            char equals;  // Using char to read the '=' character
+
+            // Read each line in the format key = value
+            if (iss >> key >> equals >> value && equals == '=' && key == "base_min_p") {
+                *base_min_p = value;
+            }
+            // Note: If the key doesn't match, or the format is incorrect, 
+            // you might want to handle the error or set a default value
+        }
+        infile.close();
+    }
+}
+
 void llama_sample_top_p(struct llama_context * ctx, llama_token_data_array * candidates, float p, size_t min_keep) {
+    // Variables for the special mode
+    float base_min_p;  // This will hold the base minimum probability value
+    float multiplied_min_p;  // This will hold the adjusted minimum probability threshold
+
+    // If p is 1.0, we switch to a different sampling mode.
     if (p >= 1.0f) {
+        printf("returning, top p set to 1.0");
         return;
     }
 
+    // If p is ~0.02, we switch to the Min P sampler
+    if (p >= 0.01f && p <= 0.03f) {
+        printf("USING MIN P SAMPLING MODE\n");
+
+        // Ensure the probabilities are calculated.
+        llama_sample_softmax(ctx, candidates);
+
+        // Print the top tokens before filtering
+        printf("Top tokens before filtering:\n");
+        for (size_t i = 0; i < candidates->size && i < 10; ++i) {
+            printf("Token %zu: %.6f%%\n", i + 1, candidates->data[i].p * 100);  // Multiplying by 100 to convert to percentage
+        }
+
+        base_min_p = 0.05;  // For example, 5% as the base minimum probability before reading the text value
+
+        read_or_write_base_min_p(&base_min_p);
```

Comment:
Is the file just used to store a single state value? If so, you can just do this like the Mirostat samplers - they store `mu` state between calls (and it has an initial value as well).

Umm, from the `llama.cpp` side, if I remember correctly I think you just have the sampler take a `float *`. Other stuff like setting the option, storing the state to be passed to the sampler when it's used, etc is handled in `common/sampling.cpp` (and associated header file).